Fixes for rapido for systemd-233 (TW), and for the nvme-rdma scenario

mwilck opened the pull request:
These fixes make the "mpath_local" scenario work for Tumbleweed.

mwilck committed:
vm.sh: use sytemd.unit=rescue.target

"systemd.unit=emergency" does not work with systemd-233 (at least).
systemd tries to isolate "initrd.target" instead, and fails because
of missing or wrong root= boot parameter. "emergency.target" boots,
but doesn't bring up the needed services.

Using "rescue.target" fixes this. This makes rapido usable on
openSUSE Tumbleweed.

mwilck commented:
Added some more stuff that qualifies as "fix".

mwilck committed:
cut_nvme_rdma.sh: add crypto/crc32.ko module

This requires extra diligence because the module name may be
ambiguous for some kernels, there's lib/crc32.ko and crypto/crc32.ko.
The latter is needed by the rdma_rxe module.

Under openSUSE/SLES, this requires a dracut patch, too
("dracut-init.sh: ignore crc32.ko in builtin test").

mwilck committed:
nvme_rdma_autorun.sh: fix dynamic debugging

Failure to activate dyndbg should not cause a fatal error.
Also, all modules must be safely loaded before activating dyndbg.

mwilck committed:
nvme_rdma_autorun.sh: dynamic debug exceptions

Some functions flood the log just too much. Deactivate them.
Something like this could become generic functionality in a later
rapido release.

mwilck committed:
nvme_rdma_autorun.sh: Fix startup message

The grep expression was broken.

mwilck committed:
Consistent MTU handling for bridges and interfaces

Let the user configure BR_MTU. For most applications, we want
it to be 9000. It's important to set this on the bridge device,
too. And the qdisc mode also has to be set on the bridge.

mwilck committed:
readme.txt: Notes on additional packages

mwilck committed:
vm.sh: set rd.neednet if necessary

dracut without systemd will not set up the network unless
rd.neednet=1 is set.

mwilck committed:
cut_nvme_rdma.sh: workaround for dracut's rootok test

When started without systemd, but with network, dracut doesn't
start all necessary services unless "$root" and "$rootok"
are set after parsing the cmdline hooks (in particular, the
network isn't brought up). Therefore, fake an "ok" root parameter
here to trick dracut into continuing. It does no harm because
we don't want to mount root anyway.

mwilck commented:
The patches 3421744 and 94dcf23 fix the bridging setup. This was necessary for me to run the nvme_rdma scenario. The other added patches (github lists them out-of-order here, see https://github.com/mwilck/rapido/commits/fix_mpath) are all related to the nvme_rdma scenario which I finally managed to run after applying all these patches.

@morbidrsa, as this touches NVMe heavily, please have a look at it, too.

ddiss commented on tools/br_setup.sh:
```diff
@@ -59,9 +59,12 @@ unwind="ip link set dev $TAP_DEV1 down; ${unwind}"
+		--pid-file=/var/run/rapido-dnsmasq-$$.pid \
+		--bind-interfaces \
 		--interface="$BR_DEV" \
+		--except-interface=lo \
```
IIUC, "--except-interface=lo" would only have an effect here if "lo" is listed in $BR_DEV.

mwilck replied:
No, "lo" is grabbed automatically unless it's either listed as except-interface, or --listen-address is used instead of --interface. From **dnsmasq(5)**:

> "Dnsmasq automatically adds the loopback (local) interface to the list of interfaces to use when the --interface option  is used"
> "Note that if no --interface option is given, but --listen-address is, dnsmasq will not automatically listen on the loopback interface"

ddiss commented on cut_nvme_local.sh:
```diff
@@ -28,3 +28,5 @@ _rt_require_lib "libkeyutils.so.1"
 	--modules "bash base" \
 	$DRACUT_EXTRA_ARGS \
 	$DRACUT_OUT
```
Will squash a change in here to _fail immediately on dracut failure. Otherwise this looks fine.

ddiss commented on cut_mpath_local.sh:
```diff
@@ -34,7 +34,7 @@ _rt_require_dracut_args
 	--include "$RAPIDO_DIR/rapido.conf" "/rapido.conf" \
 	--include "$RAPIDO_DIR/vm_autorun.env" "/vm_autorun.env" \
-	--add-drivers "virtio_scsi" \
+	--add-drivers "virtio_scsi virtio_pci sd_mod" \
```
kernel/vanilla_config includes CONFIG_BLK_DEV_SD=y and CONFIG_VIRTIO_PCI=y, so these don't really need to be listed. Won't hurt to have them though.

mwilck replied:
I'm working with stock SLE configs, where both are modules. If that's wrong, please mention it in the README :-)

mwilck commented:
Besides the rapido patches listed here, the nvme_rdma test also needs a dracut patch: danimo/dracut#4. I've tested this on Leap 42.2 and TW, with a recent SLE12 SP2 kernel.

mwilck committed:
sle12sp3_config: enabled nvme, rdma, mpath, alua

... but only add stuff we currently really need.

ddiss commented:
@mwilck I've picked up quite a few of you're changes, and queued them alongside my upstream PR at https://github.com/rapido-linux/rapido/pull/32 . I'm still working my way though the others - I plan on setting up a tumbleweed system, so that I can try to reproduce some of the issues that you're seeing.

mwilck committed:
vanilla_config: enable rdma/rxe, mpath, alua

... and ORC unwinder.

mwilck commented:
Just rebased my branch to ddiss/misc (651a470)

morbidrsa reviewed (approved):
Looks good to me

morbidrsa reviewed (commented):
Do we really need to drag in insmod, can't kmod handle it by itself?

morbidrsa reviewed (approved):
Looks good

morbidrsa reviewed (approved):
Looks good

morbidrsa reviewed (commented):
Why do you enable balooning as well? You should at least state this in the commit message.

mwilck commented:
> Do we really need to drag in insmod, can't kmod handle it by itself?

insmod will be a symlink to kmod. But AFAICS it needs to be called as "insmod" in order to resolve the `crc32.ko` module correctly. "modprobe" fails here because it determines that `crc32.ko` (the other module by the same name)  is built-in, and therefore does nothing.

mwilck commented:
> Why do you enable balooning as well? You should at least state this in the commit message.

Hm, this must have happened indirectly via some dependency, or by mistake. I'll need to double-check. I have to admit that I enabled various options (starting from the current rapido default configuration) via trial-and-error until I found a config that made the test case work. I quit at the point where that succeeded.

mwilck commented:
Updated the sle12s3 and vanilla configs again. The changes are now fairly small. This is the minimum required to run the nvme and mpath scenarios with both kernels.
Well, I added the ORC unwinder too after listening to jslaby's talk on labsconf.

mwilck commented:
Btw I reworked https://github.com/mwilck/rapido/commits/enhancements, too.

ddiss commented:
Some of the changes in this series made it upstream, but I've unfortunately lost track of the ones that didn't. @mwilck If you're still interested in getting them in, please rebase + resubmit against the upstream repo at https://github.com/rapido-linux/rapido.git .

mwilck commented:
@ddiss, that rebase and resubmit has been on my todo list for some time. My code base has also grown in the meantime I fear :-/

ddiss commented:
@mwilck no worries, happy to look at anything big or small that you'd like to get upstream.
